src/hotspot/share/runtime/unloadableMethodHandle.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_UNLOADABLE_METHOD_HANDLE_HPP
+#define SHARE_RUNTIME_UNLOADABLE_METHOD_HANDLE_HPP
+
+#include "oops/klass.hpp"
+#include "oops/oopHandle.hpp"
+#include "oops/weakHandle.hpp"
+
+// Unloadable method handle.
+//
+// Useful when one needs to hold to Method* without delaying class unloading.
+//
+// This handle can be in 3 states:
+//  1. Initial weak state. Relevant Method* is only weakly-reachable, can be cleared
+//     by class unloading.
+//  2. Accessible strong state. Relevant Method* is strongly reachable, cannot be
+//     cleared by class unloading.
+//  3. Final released state. Relevant Method* is in unknown state, and cannot be
+//     accessed.
+//
```

Comment:
Please, elaborate what state transitions are supported. Currently, my understanding is there are 3 transitions and 4 states:
* 1 -> 2
* 2 -> 3 (terminal)
* 1 -> 3 (terminal)
* 0 (empty, terminal)

Reply:
I added class-level docs for this handle, see if it reads well?

Reply:
Looks good. Thanks.